### PR TITLE
Refactor code around changing case summary

### PR DESCRIPTION
**Pull request by slorek**
https://trello.com/c/K691hCVy/704-refactor-callbacks-when-updating-case-summary

## Description
This implements the following improvements:

* Use of new form object to perform validation on case summary form submissions
* Use of service object to control the updating of the summary, creation of audit activity and sending notification emails, instead of model callbacks which relied on `User.current`
* Store new and old case summary in the activity metadata. Old activity entries migrated to the new metadata approach (albeit without the old summary, which is not used currently but could potentially be reconstructed if needed in the future)
* Improved test coverage.
* More consistent naming of the field on error messages (previously the field was referred to as the description rather than the summary, because of the model attribute name)
* Separation of GET/PATCH controller actions. We could consider extracting these to a new `summary` subresource e.g. `GET /investigation/id/summary/edit` and `PATCH /investigation/id/summary`.
* Use of the new form builder helpers to address a few UI issues
* Bonus removal of another of the few remaining usage of `User.current`

Interestingly the existing actions unusually implemented JSON responses so I have retained these in line with our ambition to expand this support.

## Checklist:
- [x] Have you documented your changes in the pull request description?
- [ ] Does the change present any security considerations?
- [ ] Is any gem functionality overloaded? Eg: Devise controller methods being overloaded.
- [ ] Has acceptance criteria been tested by a peer?
- [ ] Has the CHANGELOG been updated? (If change is worth telling users about.)

### CODE REVIEW

File: app/controllers/investigations_controller.rb [hunk 1]
```diff
@@ -75,10 +75,36 @@ def visibility
...
   def edit_summary
+    @investigation = Investigation.includes(:teams_with_edit_access).find_by!(pretty_id: params[:pretty_id]).decorate
     authorize @investigation, :update?
-    update
+
+    @form = ChangeCaseSummaryForm.new(summary: @investigation.object.description)
+  end
+
+  # PATCH /cases/1/edit_summary
+  def update_summary
```
[the remark this answers is not in the document]
Comment (slorek) [reply app/controllers/investigations_controller.rb:86]: Yes, agreed as mentioned in the description.

File: app/controllers/investigations_controller.rb [hunk 2]
```diff
@@ -75,10 +75,36 @@ def visibility
...
+    @investigation = Investigation.includes(:teams_with_edit_access).find_by!(pretty_id: params[:pretty_id]).decorate
+    authorize @investigation, :update?
+
+    @form = ChangeCaseSummaryForm.new(params.require(:change_case_summary_form).permit(:summary))
+
+    unless @form.valid?
+      return respond_to do |format|
+        format.html { render :edit_summary, status: :unprocessable_entity }
+        format.json { render json: @form.errors, status: :unprocessable_entity }
```
[the remark this answers is not in the document]
Comment (slorek) [reply app/controllers/investigations_controller.rb:95]: Yes we should, preferably with JSON schemas and Swagger docs. But it would be a story/epic in its own right. This is existing functionality.

File: config/routes.rb [hunk 3]
```diff
@@ -104,7 +104,7 @@
       get :visibility
       patch :visibility
       get :edit_summary
-      patch :edit_summary
+      patch :update_summary
```
[the remark this answers is not in the document]
Comment (slorek) [reply config/routes.rb:107]: Yes agreed, see description.

File: spec/services/change_case_summary_spec.rb [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+      end
+
+      it "changes the case summary" do
+        expect { result }.to change(investigation, :description).from(old_summary).to(new_summary)
+      end
+
+      it "creates an audit activity for summary changed", :aggregate_failures do
+        expect { result }.to change(Activity, :count).by(1)
+        activity = investigation.reload.activities.first
+        expect(activity).to be_a(AuditActivity::Investigation::UpdateSummary)
```
[the remark this answers is not in the document]
Comment (slorek) [reply spec/services/change_case_summary_spec.rb:73]: 🤔 I think this would be slightly less useful as currently we ensure that the activity is the last one returned. This change could potentially return a previous activity entry.

[the remark this answers is not in the document]
Comment (StupidCodeFactory) [reply spec/services/change_case_summary_spec.rb:73]: yes I've used `investigation.activities.find_by!(type: AuditActivity::Investigation::UpdateSummary).order(created_at: :desc)` if I remember well

[the remark this answers is not in the document]
Comment (StupidCodeFactory) [reply spec/services/change_case_summary_spec.rb:73]: you'll then need to to change `investigation.reload.activities.first` to `investigation.reload.activities.last` unless there is a default order scope

File: spec/forms/change_case_summary_form_spec.rb [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        errors.each do |property, message|
+          expect(form.errors.full_messages_for(property)).to eq([message])
+        end
+      end
+    end
+
+    context "when no summary is supplied" do
+      let(:summary) { nil }
+
+      include_examples "invalid form", [:summary, "Summary cannot be blank"]
```
Comment (frankieroberto) [inline spec/forms/change_case_summary_form_spec.rb:34]: I think the error message should be something like `"Enter the case summary"` to be consistent with the others?

File: spec/forms/change_case_summary_form_spec.rb [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    end
+
+    context "when summary is too long" do
+      let(:summary) { rand(36**10_001).to_s(36) }
+
+      include_examples "invalid form", [:summary, "Summary is too long (maximum is 10000 characters)"]
```
Comment (frankieroberto) [inline spec/forms/change_case_summary_form_spec.rb:40]: The Design System [suggests](https://design-system.service.gov.uk/components/text-input/#error-messages) an error message in the format of `"Summary must be 10,000 characters or fewer"`

File: spec/forms/change_case_summary_form_spec.rb [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+      it "populates an error message" do
+        form.validate
+        errors.each do |property, message|
+          expect(form.errors.full_messages_for(property)).to eq([message])
+        end
+      end
+    end
+
+    context "when no summary is supplied" do
+      let(:summary) { nil }
```
Comment (frankieroberto) [inline spec/forms/change_case_summary_form_spec.rb:32]: Might be more realistic test a whitespace-only string?

```suggestion
      let(:summary) { " " }
```

File: app/forms/change_case_summary_form.rb [hunk 8]
```diff
@@ -0,0 +1,9 @@
+class ChangeCaseSummaryForm
+  include ActiveModel::Model
+  include ActiveModel::Attributes
+
+  attribute :summary
+
+  validates_presence_of :summary
+  validates :summary, length: { maximum: 10_000 }
```
Comment (frankieroberto) [inline app/forms/change_case_summary_form.rb:8]: Minor suggestion:

```suggestion
  validates :summary, presence: true, length: { maximum: 10_000 }
```

File: app/services/change_case_summary.rb [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+      create_audit_activity_for_case_summary_changed
+    end
+
+    send_notification_email
+  end
+
+private
+
+  def create_audit_activity_for_case_summary_changed
+    metadata = activity_class.build_metadata(summary, old_summary)
```
[the remark this answers is not in the document]
Comment (slorek) [reply app/services/change_case_summary.rb:27]: Seems like this can't be used because the object is already saved by the time this is called.

File: spec/services/change_case_summary_spec.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (slorek) [reply spec/services/change_case_summary_spec.rb:73]: This doesn't guarantee that the last activity entry (including all other types) is the type expected though.

File: app/models/audit_activity/investigation/update_summary.rb [hunk 10]
```diff
@@ -1,10 +1,31 @@
...
 
-  def email_update_text(viewer = nil)
-    "#{investigation.case_type.upcase_first} summary was updated by #{source&.show(viewer)}."
+  def self.build_metadata(new_summary, old_summary)
+    {
+      summary: {
+        old: old_summary,
+        new: new_summary
+      }
+    }
```
[the remark this answers is not in the document]
Comment (slorek) [reply app/models/audit_activity/investigation/update_summary.rb:12]: OK thanks. Looks like we have a few different patterns going on in the various activity.

File: app/services/change_case_summary.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (slorek) [reply app/services/change_case_summary.rb:27]: I used `previous_changes` instead.

**Review by StupidCodeFactory — APPROVED** [review]
